keras_retinanet/preprocessing/generator.py

```diff
@@ -112,19 +116,32 @@ def resize_image(self, image):
     def preprocess_image(self, image):
         return preprocess_image(image)
 
-    def preprocess_group(self, image_group, annotations_group):
-        for index, (image, annotations) in enumerate(zip(image_group, annotations_group)):
-            # preprocess the image (subtract imagenet mean)
-            image = self.preprocess_image(image)
+    def preprocess_group_entry(self, image, annotations):
+        # preprocess the image
+        image = self.preprocess_image(image)
+
+        # randomly transform both image and annotations
+        if self.transform_generator is not None:
+            transform = adjust_transform_for_image(next(self.transform_generator), image)
+            apply_transform(transform, image, self.transform_parameters)
 
-            # randomly transform both image and annotations
-            image, annotations = random_transform(image, annotations, self.image_data_generator)
+            # Transform the bounding boxes in the annotations.
+            annotations = annotations.copy()
+            for index in range(annotations.shape[0]):
+                annotations[index, :4] = transform_aabb(transform, annotations[index, :4])
```

Comment:
One thing I was wondering: What should the generator do if `transform_aabb` puts the annotation outside of the image canvas? I'm assuming they will automatically be filtered now by `compute_input_output`. But maybe I overlooked something?

Reply:
Also, is that desirable? When the aabb ends up outside the image and it is filtered, that means that part of the image becomes background, even though it was actually an annotation.

Should we instead discard the entire image? Or should we try up to X times to get a transformation that doesn't put any annotations outside of the image (and discard if that failed X times)? Or should we clip the new bounding box with the image canvas?

Reply:
My first thought is to clip the annotation to the image canvas, to preserve as much of the annotations as possible. What do you think ?

Reply:
I'm not sure. I have a slight feeling that it might reduce the quality of training in some cases. Maybe some very distinctive part of the thing to recognize is clipped off, or maybe the clipped thing looks more like a different object (contrived example: a face with a clown nose vs a face with a normal nose).

Reply:
I think filtering it out entirely is more harmful then clipping. So either we make sure it doesn't occur, or we clip it in my opinion.